src/celeritas/user/DetectorSteps.cu

```diff
@@ -0,0 +1,155 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/user/DetectorSteps.cu
+//---------------------------------------------------------------------------//
+#include "DetectorSteps.hh"
+
+#include <thrust/copy.h>
+#include <thrust/count.h>
+#include <thrust/device_ptr.h>
+#include <thrust/device_vector.h>
+#include <thrust/execution_policy.h>
+#include <thrust/partition.h>
+
+#include "corecel/data/Collection.hh"
+
+#include "StepData.hh"
+
+using thrust::device_pointer_cast;
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+using DVecDetector = thrust::device_vector<DetectorId>;
+
+template<class T>
+using StateRef
+    = celeritas::StateCollection<T, Ownership::reference, MemSpace::device>;
+
+//---------------------------------------------------------------------------//
+struct HasDetector
+{
+    CELER_FORCEINLINE_FUNCTION bool operator()(const DetectorId& d)
+    {
+        return static_cast<bool>(d);
+    }
+};
+
+//---------------------------------------------------------------------------//
+size_type count_num_valid(const DVecDetector& orig_ids)
+{
+    return thrust::count_if(
+        thrust::device, orig_ids.begin(), orig_ids.end(), HasDetector{});
+}
+
+//---------------------------------------------------------------------------//
+// Simplify opaque ID pointers to reduce code bloat from identical
+// instantiations
+template<class T>
+struct PointerTransformer
+{
+    constexpr T* operator()(T* v) const { return v; }
+};
+
+template<class V, class S>
+struct PointerTransformer<OpaqueId<V, S>>
+{
+    constexpr S* operator()(OpaqueId<V, S>* v) const
+    {
+        return reinterpret_cast<S*>(v);
+    }
+};
+
+//---------------------------------------------------------------------------//
+template<class T>
+void assign_field(std::vector<T>*     dst,
+                  const StateRef<T>&  src,
+                  const DVecDetector& orig_ids,
+                  size_type           size)
+{
+    if (src.empty())
+    {
+        // This attribute is not in use
+        dst->clear();
+        return;
+    }
+
+    PointerTransformer<T> transform_ptr;
+
+    // Partition based on detector validity
+    auto temp_span = src[AllItems<T>{}];
+    thrust::partition(thrust::device,
```

Comment:
Yeah this is a question I wanted to ask @tmdelellis. The section of code here is based on shift, and I assume it works because the ordering is consistent across calls? Even if it's doesn't preserve ordering among elements with equal detector ids, it's (presumably?) doing the same operations to all the data.

Reply:
OK after further discussion with Tom and Steve:
- `stable_partition` is sufficient but not necessary
- we can't think of any reason why `partition` would have nondeterministic behavior with the same stencil
- the safest way would be to generate a steering/indirection vector with `linear_sequence`, then partition that, then use `gather`

So I think we can just stick with this for now...

Reply:
Alright, sounds reasonable!